Map position parameters for /imagery-search URL

tombh opened the pull request:
Note: the optimal event to listen for changes is `moveend` but the Mapbox
Synch plugin seems to continuously trigger it. So if the code to update
the URL is also bound to `moveend` then there is significant excess code
execution -- browser slow down and URL constantly updates. Fortunately
falling back to using a combination of `dragend` and `zoomend` seems
sufficient.

Fixes #87

tombh committed:
Track imagery-search's map view in URL for sharing

The optimal event to listen for changes is 'moveend' but the Mapbox
Synch plugin seems to continuously trigger it. So if the code to update
the URL is also bound to 'moveend' then there is significant excess code
execution -- browser slow down and URL constantly updates. Fortunately
falling back to using a combination of 'dragend' and 'zoomend' seems
sufficient.

Fixes #87

Add throttle to onMapChange()

When using `moveend` to trigger URL updates then you suffer from
continuous triggering from the map sync plugin. So using a bit of
throttling on the callback prevents jitter. However there is now an
artifact where the mouse cursor flashes between grab and cursor.
Nevertheless this currently seems like the most practical and effective
compromise.

tombh commented:
As @smit1678 pointed, zooming seems to have a bit of jitter/perf issues. Will see if what I can do...

tombh commented:
So reverting to listen to `moveend` fixes the zoom problem. However, I could only test that by disabling map syncing -- so basically the 2 things don't play well together :/

There may be [another event](https://github.com/mapbox/mapbox-gl-js/blob/95b0ca42bf8376561db7e96c616d76696d8f65f7/debug/events.html#L169) we can use. Eg; `mouseup` works well for dragging but not zooming. Or maybe there's some kind of other bad interaction going on when *all* maps are listening to `dragend` and `zoomend`. It which case a somewhat hacky solution would be to only set listeners once a `touchstart` or `mousedown` happens *inside a specific map*. But it just seems inelegant and feels like it will lead to unexpected behaviour and difficult to maintain code.

The best compromise appears at the moment to replace the map sync code with something that doesn't constantly sync the maps during manipulation, but rather syncs immediately *after* the manipulation has happened. Eg; you start dragging a little map and nothing happens, but as soon as you let go of the mouse, all the other maps instantly sync.

smit1678 commented:
@tombh Thanks for the write up and testing ways to improve the performance. So it seems that we're in a spot where we may need to compromise on one of the following: 

  1. Jittery zoom (through scroll and controls) but we keep the maps fully synced. 
  2. More stable, performant zooming but maps only synced after panning or zooming is complete. 

cc @danielfdsilva @nbumbarger for thoughts or insight on the compromise? or other ideas for work arounds?

danielfdsilva commented:
@smit1678 @tombh You could try using the `moveend` method with some `debounce` or `throttle`. That would solve the excess updates, while keeping everything in sync.

tombh commented:
Nice idea, so like updating the URL only once every 50ms of active movement for example?

tombh commented:
I reverted back to using the superior `moveend` event and added a 500ms throttle to the URL updater. It seems to work better except for a tiny niggle that the cursor flashes between states when the throttle resets. Does it seem to effect the UX, or is it purely aesthetic?

smit1678 commented:
👍  tested and works well. I don't see the cursor flash on Chrome or Firefox on OSX. I don't think it's a huge issue.

danielfdsilva commented on app/assets/scripts/views/imagery-search.js:
```diff
@@ -93,9 +122,15 @@ var ImagerySearch = React.createClass({
     ];
 
+    this.maps.map(
+      (map) => { map.on('moveend', this.onMapChange); }
```
Seems a bit unnecessary to add listeners to all maps. Since they're synced using the main map would probably be enough. Or is there a reason behind the approach.

tombh replied:
Oh! Yes, you're right. That happened because I was originally solving the jitter by using `dragend` and `zoomend` which I assume aren't broadcast to all other maps by the sync plugin. But now we have a different jitter fix I'll just bind to the main map.

danielfdsilva commented on app/assets/scripts/views/imagery-search.js:
```diff
@@ -130,6 +165,25 @@ var ImagerySearch = React.createClass({
+    // a scroll back to the top of the page. hashHistory is still a dependency because
+    // of the expectation that the host does not support SPA routing to index.html.
+    const path = `#/imagery-search/${this.mapViewToString()}`;
+    window.history.replaceState(null, null, path);
```
The scroll to the top is not related to hashHistory. It's an intentional plugin that fires when the url changes. This is needed for when we change to pages with a lot of content. It's implemented [here](https://github.com/hotosm/imagery-coordination/blob/develop/app/assets/scripts/main.js#L51-L53).

I don't have a better suggestion about this, besides adding a condition for that specific path to the scroll function, so this is more a fyi.

tombh replied:
Ah, I didn't know that. Ok, I'll change the comment to reflect that then.

BTW, am I right though about why `hashHistory` is used throughout rather than browser history API? Namely Github's lack of support for SPA routing.
